### PR TITLE
Type for Metadata

### PR DESCRIPTION
**Pull request by tim-speed**
This adds a type to store metadata.
This will be used for provision operations and resources initially.

Relates manifoldco/engineering#5801

### CODE REVIEW

**Comment by tim-speed** [comment]
@luizbranco I was trying to solve that by implementing the Validate method. But I see where you're coming from, I remember when the wrong types were being supplied for the feature values and there is no constraint or sufficient casting for them on the backend.

Hmm, and I can see how just allowing strings like you were suggesting yesterday would solve this :D cause if everything is a string, there's no worry! You always know what to expect.

What if we had key prefixes that enforced the type through the validate method?

"STR_TEMPLATE"
"INT_BACKUP_INTERVAL_DAYS"
"BOL_EMAIL_DIGESTS"
"OBJ_USER_PROFILE"

And then I could write Getter and HasValue methods for each type
`.HasString("Template")` `.GetString("Template")`
Could make it case insensitive too.

That way the client has to be very explicit of what it's sending. and we have a simple way to get the normalized values.

Let me know how that sounds, if it's no good I can just limit it to string and if someone wants to do something like a user profile it can be spread across multiple fields, where all values are strings:

"USER_PROFILE_NAME"
"USER_PROFILE_AGE"

File: types_test.go [hunk 1]
```diff
@@ -49,3 +50,123 @@ func TestFeatureMap(t *testing.T) {
...
+			"bulbasaur":  "one",
+			"charmander": true,
+		}
+		c := Metadata{
+			"abra":      1,
+			"bulbasaur": "TWO",
+		}
+
+		if a.Equals(b) {
+			t.Error("A and B are not equal, but are.")
```
Comment (Minivera) [inline types_test.go]: But are?

Comment (tim-speed) [reply types_test.go]: It's expected that they are not equal, but the error is that they are equal :D

Comment (Minivera) [reply types_test.go]: This looks to me like a double negation :P

Comment (tim-speed) [reply types_test.go]: It is! :) cause its a negative test case

**Comment by tim-speed** [comment]
From Slack:

> luiz [1:24 PM]
> :thinking_face: I guess that would work
> 
> what I was suggesting was something like
> the product defines the metadata spec
> ```key: xxx
> type string
> 
> key: yyy
> type: int```
> we save everything as string
> and when deserializing, we use the types defined on the product (edited)
> because it's going to be tied to the product/plan, right?
> 
> tim [1:26 PM]
> yeah I like that one too, it's just more time to implement, cause I'll need to add that part as well. And then we'd need some spec for custom resources somewhere, and if it's in code then we'd still need to make code changes and deploys to update it.
> 
> luiz [1:27 PM]
> I see
> I guess prefix would work
> other option would be instead of the value being a string, it could be a `{type: x, value: y}`
> 
> tim [1:28 PM]
> ahh yeah thats a good point
> cleaner to read
> and we can enforce that Value type struct
> 
> luiz [1:29 PM]
> :+1:
> 
> tim [1:29 PM]
> I'll do that :+1: thanks!
> Gonna add this convo to the GH issue

**Comment by tim-speed** [comment]
okay okay, I'll reword this :D

File: types_test.go (hunk 1 above)
Comment (tim-speed) [reply types_test.go]: okay okay, I'll reword this :D

**Comment by tim-speed** [comment]
this also has 100% coverage, we just don't have codecov enabled :'(

File: types.go [hunk 2]
```diff
@@ -16,3 +23,185 @@ func (f FeatureMap) Equals(fm FeatureMap) bool {
...
+	m.Value = val
+	if err := m.tryCastFields(); err != nil {
+		return errors.Errorf("Could not make MetadataValue from map: %s", err.Error())
+	}
+	return nil
+}
+
+// UnmarshalJSON controls how a MetadataValue is parsed from JSON
+func (m *MetadataValue) UnmarshalJSON(data []byte) error {
+	mv := &struct {
```
[the remark this answers is not in the document]
Comment (tim-speed) [reply types.go]: It get's into an unmarshall infinite loop if you do! I'll add a comment to explain.
I also want to add GetString, GetBool, GetInt,... etc methods, that return value and err

**Pull request merged**